File: frontend/types/sentiment.ts
// Sentiment domain types
// AUTO-SYNCED with backend via openapi-typescript
// Last synced: 2026-01-08
// Source: components["schemas"]["SentimentAnalysisResponse"], etc.

// ============================================
// ENUMS / UNION TYPES
// ============================================

/**
 * Sentiment classification labels
 */
export type SentimentLabel = 
  | 'very_negative' 
  | 'negative' 
  | 'neutral' 
  | 'positive' 
  | 'very_positive';

/**
 * Sentiment trend direction
 */
export type SentimentTrendDirection = 'up' | 'down' | 'stable';

/**
 * Source platforms for social mentions
 */
export type SentimentSource = 
  | 'twitter' 
  | 'reddit' 
  | 'manual' 
  | 'news' 
  | 'instagram' 
  | 'facebook' 
  | 'youtube';

// ============================================
// SOCIAL MENTION TYPES
// ============================================

/**
 * Social mention from the API
 * Matches: components["schemas"]["MentionResponse"]
 */
export interface SocialMention {
  id: string;
  product_id: string;
  source: string;
  source_id: string | null;
  content: string;
  author: string | null;
  author_followers: number | null;
  engagement_count: number | null;
  url: string | null;
  published_at: string | null;
  collected_at: string;
  processed: boolean;
  // Sentiment fields (populated after analysis)
  sentiment_score: number | null;
  sentiment_label: SentimentLabel | null;
}

/**
 * Paginated mentions response
 */
export interface PaginatedMentions {
  items: SocialMention[];
  total: number;
  page: number;
  page_size: number;
  pages: number;
}

// ============================================
// SENTIMENT ANALYSIS TYPES
// ============================================

/**
 * Request to analyze sentiment
 * Matches: components["schemas"]["SentimentAnalysisRequest"]
 */
export interface AnalyzeRequest {
  product_id: string;
  content: string;
  source?: SentimentSource;
  author?: string | null;
  url?: string | null;
  use_ai?: boolean;  // Use Gemini 2.0 Flash for analysis
}

/**
 * Response from sentiment analysis
 * Matches: components["schemas"]["SentimentAnalysisResponse"]
 */
export interface AnalyzeResponse {
  id?: string;
  sentiment_id?: string;
  product_id?: string;
  content?: string;
  text?: string;
  source?: string;
  sentiment_score: number;
  sentiment_label: SentimentLabel;
  confidence: number;
  emotions?: {
    positive: number;
    negative: number;
    neutral: number;
  };
  topics?: string[];
  is_sarcastic?: boolean;
  ai_powered?: boolean;
  analyzed_by?: 'vader' | 'gemini' | 'hybrid';
  created_at?: string;
}

// ============================================
// SENTIMENT SUMMARY & TRENDS
// ============================================

/**
 * Sentiment data point for charts
 */
export interface SentimentDataPoint {
  timestamp: string;
  score: number;
  mention_count: number;
}

/**
 * Sentiment trend response
 * Matches: components["schemas"]["SentimentTrendResponse"]
 */
export interface SentimentTrend {
  product_id: string | null;
  period_days: number;
  current_score: number | null;
  previous_score: number | null;
  change: number | null;
  trend: SentimentTrendDirection;
  timeline: SentimentDataPoint[];
}

/**
 * Aggregated sentiment summary
 * Matches: components["schemas"]["SentimentSummaryResponse"]
 */
export interface SentimentSummary {
  product_id: string;
  average_score: number;
  total_mentions: number;
  positive_count: number;
  negative_count: number;
  neutral_count: number;
  trend: SentimentTrendDirection;
}

// ============================================
// AI STATUS
// ============================================

/**
 * AI status response
 * Matches: components["schemas"]["AIStatusResponse"]
 */
export interface AIStatusResponse {
  openai_available: boolean;
  gemini_available?: boolean;
  model: string | null;
  features: string[];
}

// ============================================
// FETCH MENTIONS REQUEST
// ============================================

/**
 * Request to fetch mentions from external sources
 * Matches: components["schemas"]["FetchMentionsRequest"]
 */
export interface FetchMentionsRequest {
  product_id: string;
  sources?: SentimentSource[];
  keywords?: string[];
  limit?: number;
}

/**
 * Response from fetch mentions
 * Matches: components["schemas"]["FetchMentionsResponse"]
 */
export interface FetchMentionsResponse {
  product_id: string;
  mentions_found: number;
  mentions_analyzed: number;
  sources_searched: string[];
  errors: string[];
}

File: backend/schemas/sentiment.py
# backend/schemas/sentiment.py
from datetime import datetime
from decimal import Decimal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

# ============== Request Schemas ==============


class SentimentAnalyzeRequest(BaseModel):
    """Analyze a single piece of text"""

    text: str = Field(..., min_length=1)
    source: str = Field(default="manual", max_length=50)
    author: str | None = None
    url: str | None = None


class SentimentBulkItem(BaseModel):
    """Single item for bulk analysis"""

    text: str = Field(..., min_length=1)
    source: str | None = None
    author: str | None = None
    url: str | None = None


class SentimentBulkRequest(BaseModel):
    """Analyze multiple texts at once"""

    items: list[SentimentBulkItem]


# ============== Response Schemas ==============


class SentimentScores(BaseModel):
    """VADER sentiment scores"""

    compound: Decimal = Field(ge=-1, le=1)
    positive: Decimal = Field(ge=0, le=1)
    negative: Decimal = Field(ge=0, le=1)
    neutral: Decimal = Field(ge=0, le=1)


class SentimentRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    product_id: UUID
    source: str
    raw_text: str
    compound_score: Decimal
    positive_score: Decimal
    negative_score: Decimal
    neutral_score: Decimal
    author: str | None
    url: str | None
    analyzed_at: datetime


class SentimentAnalyzeResponse(BaseModel):
    """Response for single text analysis"""

    text: str
    scores: SentimentScores
    label: str
    saved: bool = False
    sentiment_id: UUID | None = None


class SentimentSummary(BaseModel):
    """Aggregated sentiment for a product"""

    product_id: UUID
    total_mentions: int = 0
    total_records: int = 0
    average_compound: Decimal | None = None
    average_score: Decimal | None = None
    positive_count: int = 0
    negative_count: int = 0
    neutral_count: int = 0
    label_distribution: dict | None = None
    trend: str | None = None
    period_start: datetime | None = None
    period_end: datetime | None = None


class SentimentResponse(BaseModel):
    """Response matching route expectations"""

    model_config = ConfigDict(from_attributes=True)
    sentiment_id: UUID | None = None
    text: str | None = None
    sentiment_score: float = Field(ge=-1, le=1)
    sentiment_label: str
    confidence: float = Field(ge=0, le=1)
    emotions: dict | None = None
    # AI-powered fields
    topics: list[str] | None = None
    is_sarcastic: bool | None = None
    ai_powered: bool = False


class SocialMentionResponse(BaseModel):
    """Response for social mention data"""

    model_config = ConfigDict(from_attributes=True)
    id: UUID
    product_id: UUID
    source: str
    source_id: str | None = None
    content: str
    author: str | None = None
    author_followers: int | None = None
    engagement_count: int | None = None
    url: str | None = None
    published_at: datetime | None = None
    collected_at: datetime
    processed: bool = False


class AIStatusResponse(BaseModel):
    """Response for AI status check"""

    openai_available: bool
    model: str | None = None
    features: list[str] = []
